pytorch_lightning/loops/closure.py

```diff
@@ -0,0 +1,113 @@
+from abc import ABC, abstractmethod
+from dataclasses import dataclass
+from typing import Any, Callable, Optional
+
+from torch import Tensor
+
+from pytorch_lightning.profiler import BaseProfiler
+from pytorch_lightning.trainer.connectors.logger_connector.result import ResultCollection
+from pytorch_lightning.utilities.warnings import WarningCache
+
+
+@dataclass
+class ClosureResult:
+    closure_loss: Optional[Tensor]
+    result_collection: Optional[ResultCollection]
+
+    @property
+    def loss(self) -> Optional[Tensor]:
+        if self.closure_loss is not None:
+            return self.closure_loss.detach().clone()
+
+
+class Closure(ABC):
+    """
+    Abstract base class for optimizer closures in Lightning.
+
+    Formally, a closure is binding variables from an external scope to a function that does a computation on these
+    variables without taking them explicitly as input. This has the benefit that a closure can be passed to an
+    object which later can call it like a function but without requiring to pass in any arguments.
+
+    This class provides a simple abstraction making the instance of this class callable like a function while capturing
+    the :class:`ClosureResult` and caching it.
+    """
+
+    def __init__(self) -> None:
+        super().__init__()
+        self._result = None
+
+    @property
+    def result(self) -> ClosureResult:
+        """The cached result from the last time the closure was called."""
+        return self._result
```

Comment:
Do we really want a reference to this? Usually we only need to create the closure once and can reuse it. This might introduce some unwanted longtime references...

Reply:
> Do we really want a reference to this?
> This might introduce some unwanted longtime references...

so we definitely need a way to get the result out of the closure once it has run, therefore the reference here. This is the scenario:

```python
closure = Closure()
optimizer.step(closure)  # runs the closure, inside it will call model.training_step()
step_result = closre.result  # for further processing in our loops
```

Previously, it worked like this (pseudo-code):

```python
result = Result()  # empty result
closure = partial(loop.training_step_and_backward, result)
optimizier.step(closure)  # internally the closure will have a reference to result and call result.update(contents...). 
return result # here result contains stuff
```

> Usually we only need to create the closure once and can reuse it

So in the current design (and the previous one) the closure gets created on every step. 
This means the memory for the result here is getting freed once closure gets gc'd. 

However, if we want to build the closure object only once as you suggest, we may want to make the result property a get method that returns the result but also sets the self._reset to None. So how about this:

```python
def get_result(self):
    result = self._result
    self._result = None
    return result
```

Reply:
I like that a lot because that way we ensure to not have this reference forever. Could also be a property though, so that you can still do `closure.result` :)

Reply:
Agree there, `get_result` is definitely a good choice.